src/evo/mnhftx.cpp

```diff
@@ -296,6 +286,8 @@ CMNHFManager::Signals CMNHFManager::GetFromCache(const CBlockIndex* const pindex
     }
     if (!m_evoDb.Read(std::make_pair(DB_SIGNALS, blockHash), signals)) {
         LogPrintf("CMNHFManager::GetFromCache: failure: can't read MnEHF signals from db for %s\n", pindex->GetBlockHash().ToString());
+        // TODO: can it actually happen? Maybe if try to use old version with re-index and update after MN_RR activated
+        assert(false);
```

Comment:
why'd we add this?

Reply:
if it will happen (this line will be run), it means that instead legit list of signals the node will have an empty list (that is not expected).

probably we can throw some exception out then? not sure what is good enough solution; but current implementation behaviour is not expecting (because it can cause that some hard-fork will be deactivated or something similar)

Reply:
Well; we shouldn't be asserting in paths that are affected by p2p; which this appears to be (as I think this is called on block connect)

Reply:
@UdjinM6 can you evaluate these asserts?

Reply:
We should never reach this point on a potential invalid chain (mined past v20 activation block by some older version) because it's going to fail much earlier due to other changes activated via v20 I think. So this `assert` simply ensures that we read signals from evodb for some block only if we wrote them earlier. And if not, there is `GetFromCache` somewhere that is called too early which would be a logical error and we shouldn't continue. In fact, I would probably make it even more explicit 7d0cb4ac5c. We could probably also avoid writing useless data to evodb for pre-v20 blocks while we are at it 2ca4aeeea9.